Bootstrap enough registration functionality for embedded-hydrogen work

MidhunSureshR opened the pull request.

MidhunSureshR committed:
fetch registration flows

MidhunSureshR committed:
Implement registration endpoint

MidhunSureshR committed:
Implement dummy registration logic

MidhunSureshR committed:
Temporary fix for 401 errors

MidhunSureshR committed:
Fix errors

MidhunSureshR committed:
Pass in params to BaseRegistrationStage

MidhunSureshR committed:
Implement m.login.terms stage

MidhunSureshR committed:
Add startRegistration method

MidhunSureshR committed:
Allow register without providing username

MidhunSureshR committed:
Return username when registration is completed

MidhunSureshR committed:
Copy over username only if it exists

MidhunSureshR committed:
Specify what errors are ignored in options

MidhunSureshR committed:
Don't cache GET requests

MidhunSureshR committed:
No need to explicitly pass in inhibitLogin

MidhunSureshR committed:
Give default values to parameters

MidhunSureshR committed:
Object.assign is overkill here

MidhunSureshR committed:
Remove space

MidhunSureshR committed:
Change type to show that username is optional

MidhunSureshR committed:
REFACTOR: Inline method

MidhunSureshR committed:
firstStage should be a local variable

MidhunSureshR committed:
Remove unnecessary getters

MidhunSureshR committed:
Use typescript style that was agreed on earlier

MidhunSureshR committed:
No need for this export

MidhunSureshR committed:
Throw error instead of returning it

MidhunSureshR committed:
Add a tos getter

MidhunSureshR committed:
Fill in ts types + change names

MidhunSureshR committed:
Type RegistrationResponse

MidhunSureshR committed:
Refactor type

- Change name
- Move union type down

MidhunSureshR committed:
Move types to types.ts

MidhunSureshR committed:
Implement flow seclector

MidhunSureshR committed:
Use union of types for RegistrationResponse

MidhunSureshR committed:
Give proper names

MidhunSureshR commented on src/matrix/net/HomeServerRequest.ts:
```diff
@@ -30,13 +31,18 @@ export class HomeServerRequest implements IHomeServerRequest {
     private readonly _promise: Promise<any>;
 
-    constructor(method: string, url: string, sourceRequest: RequestResult, log?: ILogItem) {
+    constructor(method: string, url: string, sourceRequest: RequestResult, options?: IRequestOptions) {
```
`IRequestOptions` are the options for the `HomeServerAPI._requestFn`, should we be using it here?

bwindels replied:
No, the options passed to the different methods on HomeServerApi is actually a different object, the typing seems wrong. Well spotted. This object has log, timeout, uploadProgress and prefix (which we should not put on there really, but should just be an argument of _get, _post, ...). At the same time, `IRequestOptions` does not have log or prefix. Can you add a type and fix it? :pray:

bwindels commented on src/matrix/net/HomeServerRequest.ts:
```diff
@@ -30,13 +31,18 @@ export class HomeServerRequest implements IHomeServerRequest {
             log?.set("status", response.status);
             // ok?
-            if (response.status >= 200 && response.status < 300) {
+            if (response.status >= 200 && response.status < 300 || options?.allowedErrors?.find(e => e === response.status)) {
```
```suggestion
            if (response.status >= 200 && response.status < 300 || options?.allowedErrors?.includes(response.status)) {
```

bwindels commented on src/matrix/registration/Registration.ts:
```diff
@@ -0,0 +1,69 @@
+    constructor(hsApi: HomeServerApi, accountDetails: AccountDetails, flowSelector?: FlowSelector) {
+        this._hsApi = hsApi;
+        this._accountDetails = accountDetails;
+        this._flowSelector = flowSelector ?? (flows => flows.pop());
```
best not to modify the argument? can be confusing during debugging.
```suggestion
        this._flowSelector = flowSelector ?? (flows => flows[flows.length - 1]);
```

bwindels commented on src/matrix/registration/types/types.ts:
```diff
@@ -0,0 +1,58 @@
+/*
```
why not put this file directly in `registration`?

bwindels commented on src/matrix/net/HomeServerApi.ts:
```diff
@@ -163,6 +156,21 @@ export class HomeServerApi {
     }
 
+    register(username: string | null, password: string, initialDeviceDisplayName: string, auth?: Record<string, any>, inhibitLogin: boolean = true , options: IRequestOptions = {}): IHomeServerRequest {
+        Object.assign(options, { allowedErrors: [401] });
```
```suggestion
        options.allowedErrors = [401];
```

bwindels commented on src/matrix/registration/stages/BaseRegistrationStage.ts:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    parseResponse(response: RegistrationResponse) {
+        if ("user_id" in response) {
```
probably better to check the status code being 200 here rather than look for properties that happen to only be there when we're done?

MidhunSureshR replied:
RegistrationResponse is a union of different types so we need some way of [narrowing the type](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#the-in-operator-narrowing) (otherwise typescript complains). We could check the status code and then narrow the type explicitly but where do we get the status code?
We could add a `statusCode` method to HomeServerRequest.

bwindels commented on src/matrix/registration/stages/BaseRegistrationStage.ts:
```diff
@@ -0,0 +1,61 @@
+    protected _nextStage: BaseRegistrationStage;
+    protected _params?: Record<string, any>
+
+    constructor(hsApi: HomeServerApi, accountDetails: AccountDetails, session: string, params?: RegistrationParams) {
```
I would personally move doing the actual /register request to the `Registration` class, and make the `*Stage` classes only responsible for creating the next `AuthenticationData` to send. This way, `Registration` would have a `submitStage` method that does the register request with the auth data of the stage?

bwindels commented on src/matrix/registration/stages/BaseRegistrationStage.ts:
```diff
@@ -0,0 +1,61 @@
+     * - the next stage if this stage was completed successfully
+     * - user-id (string) if registration is completed
+     */
+    abstract complete(auth?: AuthenticationData): Promise<BaseRegistrationStage | string>;
```
why not store the access token, user name, ... on the Registration once were done and just return undefined to indicate that we're done? `BaseRegistrationStage | string` feels like it's a bit of a confusing API.

Taking the `submitStage` suggestion from below into account, this would actually become the return type of that method. I imagine you'd write something like this to consume this API:

```js
const registration = ...;
let stage = await registration.start();
while (stage) {
  switch (stage.type) {
   case ...:
       stage.setSomething(...);
       break;
 }
  stage = await registration.submitStage(stage);
}
return registration.sessionInfo; // contains homeServer, accessToken, userId, deviceId, ...
```

bwindels commented on src/matrix/registration/Registration.ts:
```diff
@@ -0,0 +1,69 @@
+        return this.parseStagesFromResponse(response);
+    }
+
+    parseStagesFromResponse(response: RegistrationResponseMoreDataNeeded): BaseRegistrationStage {
```
```suggestion
    private parseStagesFromResponse(response: RegistrationResponseMoreDataNeeded): BaseRegistrationStage {
```

bwindels replied:
After refactoring, can you also make other remaining methods that don't need to be public private?
Non-public methods are methods SDK users can't use and hence complain about when we change them, so less maintenance burden.

bwindels reviewed (commented):
great start, some suggested changes to simplify things.

MidhunSureshR committed:
Use includes instead of elaborate find

Co-authored-by: Bruno Windels <brunow@matrix.org>

MidhunSureshR committed:
Don't mutate flows

MidhunSureshR committed:
Remove type directory

MidhunSureshR committed:
No need for Object.assign here either

Co-authored-by: Bruno Windels <brunow@matrix.org>

MidhunSureshR committed:
Refactor code

- Move all code that does /register to Registration.ts
- RegistrationStage only deals with the generation of auth data
- Change API so that undefined is returned instead of string when
  registration is over

MidhunSureshR committed:
completed is not always present

MidhunSureshR committed:
Check response code instead of existence of props

MidhunSureshR committed:
Fix incorrect types

MidhunSureshR committed:
Make methods private + some props readonly

MidhunSureshR committed:
Implement responseCode()

MidhunSureshR commented on src/matrix/net/RequestScheduler.ts:
```diff
@@ -50,6 +50,13 @@ class Request implements IHomeServerRequest {
+        const promise: Promise<number> = new Promise(r => resolve = r);
+        this.requestResult?.responseCode().then(code => resolve(code));
+        return promise;
+    }
```
I'm not sure if this is what I'm supposed to do.. wdyt?

bwindels replied:
Let's discuss OOB

bwindels commented on src/matrix/net/HomeServerApi.ts:
```diff
@@ -133,37 +134,52 @@ export class HomeServerApi {
 
-    passwordLogin(username: string, password: string, initialDeviceDisplayName: string, options?: IRequestOptions): IHomeServerRequest {
+    register(username: string | null, password: string, initialDeviceDisplayName: string, auth?: Record<string, any>, inhibitLogin: boolean = true , options: BaseRequestOptions = {}): IHomeServerRequest {
+        options.allowedErrors = [401];
```
I'd put this at the call site in Registration I think, as that's where we'll actually handle the 401.

bwindels commented on src/matrix/registration/types.ts:
```diff
@@ -0,0 +1,61 @@
+    status: 401;
+} 
+
+type RegistrationResponseError = {
```
I don't think you'll ever receive a response like this from HomeServerApi as the only allowedError is currently 401.
Probably best to use the existing HomeServerError for this though, as both the errcode and error are available on that?

bwindels commented on src/matrix/registration/stages/BaseRegistrationStage.ts:
```diff
@@ -0,0 +1,52 @@
+    protected _nextStage: BaseRegistrationStage;
+    protected readonly _params?: Record<string, any>
+
+    constructor(hsApi: HomeServerApi, accountDetails: AccountDetails, session: string, params?: RegistrationParams) {
```
I think we don't need hsApi or accountDetails here anymore, do we?

bwindels commented on src/matrix/registration/registrationStageFromType.ts:
```diff
@@ -0,0 +1,32 @@
+
+type ClassDerivedFromBaseRegistration = { new(hsApi: HomeServerApi, registrationData: AccountDetails, session: string, params?: RegistrationParams): BaseRegistrationStage } & typeof BaseRegistrationStage;
+
+export function registrationStageFromType(type: string): ClassDerivedFromBaseRegistration | undefined{
```
A nit, but I would just put this in registration and return an instance immediately rather than the constructor?

bwindels commented on src/matrix/registration/Registration.ts:
```diff
@@ -0,0 +1,114 @@
+            throw new Error("flowSelector did not return any flow!");
+        }
+        let firstStage: BaseRegistrationStage | undefined;
+        let lastStage: BaseRegistrationStage;
```
```suggestion
        let lastStage: BaseRegistrationStage | undefined;
```

bwindels reviewed (commented):
Thanks for the changes, some final thoughts but looking good.

bwindels commented on src/matrix/registration/stages/BaseRegistrationStage.ts:
```diff
@@ -0,0 +1,52 @@
+     * This method should return auth part that must be provided to
+     * /register endpoint to successfully complete this stage
+     */
+    abstract generateAuthenticationData(): AuthenticationData;
```
Forgot before, but we should mark methods that are not meant to be used outside of the SDK as `@internal`, we can then prevent these from being put in the d.ts file with the [`stripInternal`](https://www.typescriptlang.org/tsconfig#stripInternal) option. 
```suggestion
    /** @internal */
    abstract generateAuthenticationData(): AuthenticationData;
```

MidhunSureshR committed:
Fix responseCode in Request

MidhunSureshR committed:
Make more changes

- make setter a method
- lazily create promise

MidhunSureshR committed:
Rename allowerErrors -> allowedStatusCodes

MidhunSureshR committed:
Remove error code

MidhunSureshR committed:
Remove dead code

MidhunSureshR committed:
Create registration stage in Registration itself

MidhunSureshR committed:
Let type also be undefined

MidhunSureshR committed:
mark method as internal

Co-authored-by: Bruno Windels <bruno@windels.cloud>

MidhunSureshR committed:
Merge branch 'registration' of github.com:vector-im/hydrogen-web into registration

MidhunSureshR committed:
Fix params

MidhunSureshR committed:
Remove dead imports

bwindels commented on src/matrix/net/RequestScheduler.ts:
```diff
@@ -25,31 +25,58 @@ import type {IHomeServerRequest} from "./HomeServerRequest.js";
         } else {
-            this.reject(new AbortError());
+            this.responseReject(new AbortError());
+            this.responseCodeReject(new AbortError());
```
this might not be set yet, I think?

bwindels commented on src/matrix/net/RequestScheduler.ts:
```diff
@@ -113,9 +140,7 @@ export class RequestScheduler {
-                    request.requestResult = requestResult;
-                    const response = await requestResult.response();
-                    request.resolve(response);
+                    request.setRequestResult(requestResult);
```
the await was important here as it prevents tight-looping and allows to catch any errors.

MidhunSureshR committed:
responseCodeReject may not exist

MidhunSureshR committed:
Use async/await

bwindels reviewed (approved):
Thanks for the changes, looks good!
